velox/vector/fuzzer/VectorFuzzer.cpp

```diff
@@ -865,6 +867,69 @@ RowVectorPtr VectorFuzzer::fuzzRowChildrenToLazy(RowVectorPtr rowVector) {
       std::move(children));
 }
 
+// Utility function to check if a RowVector can have nested lazy children.
+bool canWrapChildrenInLazy(const VectorPtr& baseVector) {
+  if (baseVector->typeKind() == TypeKind::ROW) {
+    if (VectorEncoding::isDictionary(baseVector->encoding())) {
+      return canWrapChildrenInLazy(baseVector->valueVector());
+    }
+    RowVector* rowVector = baseVector->as<RowVector>();
+    if (rowVector) {
+      return rowVector->nulls() == nullptr;
+    }
+  }
+
+  return false;
+}
+
+// Utility function That only takes a row vector that passes the check in
+// canWrapChildrenInLazy() and either picks the first child to be wrapped in
+// lazy OR picks the first row vector child that passes canWrapChildrenInLazy()
+// and traverses its children recursively.
+VectorPtr wrapInnermostChildToLazy(VectorPtr& baseVector) {
```

Comment:
Would it work to just turn on VectorFuzzer::allowLazyVector? This allows generating lazy children recursively too.

Reply:
We need a deterministic way of generating lazy vectors and comparing them against a run in the simplified path, this means that we generate a regular vector, then randomly pick vectors to wrap in lazy to be passed to the common path, finally also keep track of which ones we wrapped in lazy so that we can repro. allowLazyVector would instead generate lazy randomly at the start, so we wont be able to do the aforementioned things.

Reply:
Wouldn't the vector fuzzer still be deterministic even if we turn on allowLazyVector as long as it use the same random number generator (i.e., rng_) in it? We can generate an input vector with random lazy child in this way, then load all lazy child before feeding it to the simplified path. Would this work?

Reply:
@bikramSingh91 Does your comment imply that there are parts of the fuzzer that are not fully deterministic - given a set seed ?

Reply:
To clarify, yes the fuzzer would deterministically generate a lazy vector, however in this case we need to pass the same top level input row vector to multiple Expression eval paths (common, simplified, with try wraps, etc) so we need to make sure the exact same input is passed. Since lazy vectors are loaded in place, if we generate a lazy via the fuzzer, it will be modified on the first eval path itself. To work around this, I generate a regular vector, then randomly pick columns/children in the top level vector (see generateLazyColumnIds()). This way I can wrap them in lazy right before passing them to every eval path. Now, in this PR I am leveraging the same columns ids I generated for wrapping in lazy and make sure wrapInnermostChildToLazy() deterministically chooses the first innermost eligible child that I can wrap in lazy

Reply:
Got it. Also I think the way ExpressionRunner reproduces lazy input vectors is through fuzzRowChildrenToLazy(rowVector, columnsToWrapInLazy) which requires us to have a method that wraps child vectors according to columnsToWrapInLazy.